### ISSUE TITLE
Debugger not working with Spanish language load

### ISSUE THREAD

**Issue by Susensio** [explosion/spaCy#2673]
## Edit:

It ends up working but it takes several minutes to get past that point.

## Original post

I found about this on VSCode, but same happens debugging with pdb.
Running the following script works fine, debugging it hangs indefinitely. It fails with Spanish language, but English loads correctly in debugger.

```python
import spacy
nlp = spacy.load('es')
x = 1 + 2
```

I have _debugged into_ the problem, and this is the most deep I have got:

```python
> /home/susensio/Projects/test.py(3)<module>()
-> nlp = spacy.load('en')
...
> /opt/anaconda/anaconda3/lib/python3.7/site-packages/spacy/__init__.py(15)load()
-> return util.load_model(name, **overrides)
...
> /opt/anaconda/anaconda3/lib/python3.7/site-packages/spacy/util.py(112)load_model()
-> return load_model_from_link(name, **overrides)
...
> /opt/anaconda/anaconda3/lib/python3.7/site-packages/spacy/util.py(129)load_model_from_link()
-> return cls.load(**overrides)
...
> /opt/anaconda/anaconda3/lib/python3.7/site-packages/spacy/data/es/__init__.py(12)load()
-> return load_model_from_init_py(__file__, **overrides)
...
> /opt/anaconda/anaconda3/lib/python3.7/site-packages/spacy/util.py(173)load_model_from_init_py()
-> return load_model_from_path(data_path, meta, **overrides)
...
> /opt/anaconda/anaconda3/lib/python3.7/site-packages/spacy/util.py(143)load_model_from_path()
-> cls = get_lang_class(meta['lang'])
...
> /opt/anaconda/anaconda3/lib/python3.7/site-packages/spacy/util.py(50)get_lang_class()
-> module = importlib.import_module('.lang.%s' % lang, 'spacy')
...
> /opt/anaconda/anaconda3/lib/python3.7/importlib/__init__.py(127)import_module()
-> return _bootstrap._gcd_import(name[level:], package, level)
...
(Pdb) n
> <frozen importlib._bootstrap>(1006)_gcd_import()
(Pdb) n
```

I cannot get past this point...

I'm running all this from conda

## Info about spaCy

* **spaCy version:** 2.0.11
* **Platform:** Linux-4.9.0-7-amd64-x86_64-with-debian-9.5
* **Python version:** 3.7.0
* **Models:** en, es, es_core_news_md

This may be related somehow:
#1640

**Comment by ines** [comment]
> It ends up working but it takes several minutes to get past that point.

Thanks for updating – does it mean that it all runs as expected, but takes long?

If so, one reason for this could be that the debugger reloads the model (maybe while evaluating some other stuff) *and* that the languages are lazy-loaded and now quite large (for example, Spanish ships with the lemmatizer lookup table). So depending on what the debugger does and what it evaluates, it can potentially take a while. Maybe there are some tricks we can use internally to make this faster, but I'm not sure!

**Comment by Susensio** [comment]
Well, not exactly as expected. I thing something is not working properly in some languages.

In normal executions, both English and Spanish are loaded in virtually no time (2 seconds at most).
While debugging, English takes 2 seconds but Spanish takes several minutes.

The problem seems to raise when importing spacy.lang.es
Maybe its some bug of `importlib`, but it seems curious the similarity with #1640 and that it only affects some languages...

I have localized the problem in function call `_load_unlocked` at line 967 of `importlib._bootstrap`
I'll keep digging

**Comment by Susensio** [comment]
I wasn't able to find the cause. I just resigned not to debug my app with debugging tools.
But now this has showed up somewhere else. When I run **coverage tests**, same initial delay occurs. It keeps loading spaCy model for **several minutes**, which is a bit inconvenient.
Curiously, if I just run the tests, it doesn't get stuck.

Example code that replicates this bug:

sample.py
```
import spacy

def load():
    nlp = spacy.load('es')
    x = 1 + 2
    return x

if __name__ == '__main__':
    load()
```

```
$ time python sample.py
real    0m1.547s
user    0m1.584s
sys     0m0.280s
```

```
$ time coverage run sample.py
real    3m13.773s
user    3m13.804s
sys     0m0.248s
```

**Comment by thiagoananias** [comment]
Hello! The same occurs with the pt language

**Comment by arcodergh** [comment]
I have the same issue while load the Spanish model in debug mode, is the same if I create a blank model.

**Comment by GyllingSW** [comment]
I have the same issue with both the Spanish model and a custom Danish model.

Any suggestions to what is the cause?

**Comment by ines** [comment]
Thanks for looking into this in more detail!

> `from .lemmatizer import LOOKUP  <-- This line takes forever in debug mode`

Hmmm, okay, this might explain a lot. The lemmatization tables are regular Python dictionaries and they can be quite large. So if the debugger is trying to analyse them in more detail, this can potentially be very expensive.

Maybe you could add a rule to the debugger that explicitly excludes those files? E.g. something like this: https://stackoverflow.com/a/47566410/6400719

I've actually ended up doing something similar in our `flake8` config, because running that over the lemmatization tables also isn't feasible. See here for the exclude rules: https://github.com/explosion/spaCy/blob/develop/.flake8#L5

**Comment by RafaelLeonhardt** [comment]
> 
> 
> Hello! The same occurs with the pt language
```
import spacy
import datetime

def load():
    print('ENGLISH MODEL in debug mode')
    print(datetime.datetime.now())
    nlp = spacy.load('en')
    print(datetime.datetime.now())
    print('\nPORTUGUESE MODEL in debug mode')
    print(datetime.datetime.now())
    nlp = spacy.load('pt')
    print(datetime.datetime.now())    
    x = 1 + 2
    return x

if __name__ == '__main__':
    load()
```

ENGLISH MODEL - run into 3 seconds
2019-02-18 14:**35:09.98**4418
2019-02-18 14:**35:12.54**0510

PORTUGUESE MODEL - run into 5 minutes
2019-02-18 14:**35:12.54**1511
2019-02-18 14:**40:36.80**9234

**Comment by ines** [comment]
A lot of this should have been fixed in #3218 (which also includes a nice speed comparison graph). I'm merging this with #3258, since the remaining problems really mostly come down to the size of the lemmatization tables and similar data.
